### PR TITLE
feat(compass-collection): Collection Header Experimentation Integration for Mock Data Generator – CLOUDP-333847

### PR DESCRIPTION
**Pull request by jcobis**
## Description
Implements the Mock Data Generator button in the Collection Header Actions component. The feature adds a new "Generate Mock Data" button that appears conditionally based on experiment assignment, Atlas environment, and collection type.

Also adds the modal state to the Redux store.

### Enabled
Screenshot 2025-08-08 at 3 39 24 PM

### Disabled
Screenshot 2025-08-08 at 3 41 45 PM

## Motivation and Context
This implements the Mock Data Generator feature integration into the Collection Header, allowing users to access AI-powered mock data generation directly from the collection view. The feature is gated behind an experiment.
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx [hunk 1]
```diff
@@ -81,6 +109,26 @@ const CollectionHeaderActions: React.FunctionComponent<
           Open MongoDB shell
         </Button>
       )}
+      {shouldShowMockDataButton && (
+        <Tooltip
+          enabled={!hasData}
+          trigger={
+            <div>
+              <Button
+                data-testid="collection-header-generate-mock-data"
```
Comment (kpamaran) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.tsx]: `data-testid` should have a `-button` suffix or similar

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 2]
```diff
@@ -168,4 +204,182 @@ describe('CollectionHeaderActions [Component]', function () {
       );
     });
   });
+
+  context('Mock Data Generator Button', function () {
```
Comment (kpamaran) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:200]: I suggest adding `data-testid` for the tooltip and asserting its existence directly in maybe new tests like 'should show Tooltip when ...'

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 3]
```diff
@@ -168,4 +204,182 @@ describe('CollectionHeaderActions [Component]', function () {
...
+        {},
+        atlasConnectionInfo
+      );
+
+      const button = screen.getByTestId('collection-header-generate-mock-data');
+      button.click();
+
+      expect(onOpenMockDataModal).to.have.been.calledOnce;
+    });
+  });
```
Comment (kpamaran) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:377]: would be nice to have a test that the modal opens initially with step `MockDataGeneratorStep.AI_DISCLAIMER` since that is controlled by the header

**Review by kpamaran — APPROVED** [review]
Only had minor feedback around testing. Looks good in general

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx (hunk 2 above)
Comment (jcobis) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:200]: Good idea, how about we add this once we hook up the data so we can test its presence when collection is empty?

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx (hunk 3 above)
Comment (jcobis) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:377]: Note: what step we open up to will depend on the state of project level settings, as well as a user-level "opt in" flag. We do have the test `should call onOpenMockDataModal when CTA button is clicked`, so I think the step we open to will depend based on functionality we will add in later tickets. Do you think this test is sufficient for now, or would you suggest adding something?

Comment (kpamaran) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:377]: That's good. No point in adding a test that'd be irrelevant so soon

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx (hunk 2 above)
Comment (kpamaran) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx:200]: Sounds good

File: packages/compass-telemetry/src/growth-experiments.ts [hunk 4]
```diff
@@ -2,3 +2,8 @@ export enum ExperimentTestName {
   earlyJourneyIndexesGuidance = 'EARLY_JOURNEY_INDEXES_GUIDANCE_20250328',
   mockDataGenerator = 'MOCK_DATA_GENERATOR_20251001',
 }
+
+export enum ExperimentTestGroup {
+  mockDataGeneratorVariant = 'mockDataGeneratorVariant',
+  mockDataGeneratorControl = 'mockDataGeneratorControl',
+}
```
Comment (addaleax) [inline packages/compass-telemetry/src/growth-experiments.ts:9]: Just a small note, it's not a hard-and-fast rule, but we generally have a mild preference for literal unions over enum types – i.e. `export type ExperimentTestGroup = 'mockDataGeneratorVariant' | 'mockDataGeneratorControl';` since that has a few small advantages over `enum`s:

- You don't unnecessarily alias a value to a property name if they're both the same anyway
- You can often avoid importing and resolving the module at runtime when using enum values, and instead either have a type-only import or none at all
- Literal unions would be compatible with [`--erasableSyntaxOnly`](https://www.totaltypescript.com/erasable-syntax-only) if we ever decide to adopt this flag
- Literal unions are also just generally a bit more flexible and can include additional values if we need to add them later, e.g. ``type Foo = 'abc' | 'def' | `numeric:${number}`;``

**Review by addaleax — APPROVED** [review]

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 5]
```diff
@@ -9,46 +12,83 @@ import {
...
+        </CompassExperimentationProvider>
+      </TelemetryProvider>
     );
+
+    if (connectionInfo) {
+      // For tests that need Atlas metadata (Mock Data Generator button visibility, etc.)
+      return await renderWithActiveConnection(ui, connectionInfo);
+    } else {
+      // For tests that only need basic component rendering (readonly checks, view buttons, etc.)
+      const { render } = await import('@mongodb-js/testing-library-compass');
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx]: Why is this a dynamic import?

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 6]
```diff
@@ -9,46 +12,83 @@ import {
...
+              <CollectionHeaderActions {...defaultProps} />
+            </PreferencesProvider>
+          </WorkspacesServiceProvider>
+        </CompassExperimentationProvider>
+      </TelemetryProvider>
     );
+
+    if (connectionInfo) {
+      // For tests that need Atlas metadata (Mock Data Generator button visibility, etc.)
+      return await renderWithActiveConnection(ui, connectionInfo);
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx]: You can just use it for both cases, what difference does it make?

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 7]
```diff
@@ -9,46 +12,83 @@ import {
...
-      </WorkspacesServiceProvider>
+    const defaultProps = {
+      namespace: 'test.test',
+      isReadonly: false,
+      onOpenMockDataModal: sinon.stub(),
+      ...props,
+    };
+
+    const ui = (
+      <TelemetryProvider options={{ sendTrack: sinon.stub() }}>
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx]: compass-testing-library render helpers already provide default track function, why are you adding this provider?

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx [hunk 8]
```diff
@@ -168,4 +204,184 @@ describe('CollectionHeaderActions [Component]', function () {
...
+          isReadonly: false,
+        },
+        {},
+        atlasConnectionInfo
+      );
+
+      expect(
+        screen.getByTestId('collection-header-generate-mock-data-button')
+      ).to.exist;
+      expect(screen.getByText('Generate Mock Data')).to.exist;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx]: get by text is already pretty good, get by role is even better, why do you duplicate this check with get by test id?

File: packages/compass-collection/src/components/collection-header/collection-header.tsx [hunk 9]
```diff
@@ -106,6 +108,16 @@ export const CollectionHeader: React.FunctionComponent<
   const connectionId = connectionInfo.id;
   const connectionName = getConnectionTitle(connectionInfo);
 
+  // Mock Data Generator modal state
+  const [isMockDataModalOpen, setIsMockDataModalOpen] = useState(false);
+  const [currentStep, setCurrentStep] = useState(
+    MockDataGeneratorStep.AI_DISCLAIMER
+  );
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx]: I'm not sure that having this as a UI state makes sense: you have form state transitions that clearly would depend on async actions that are handled by the store and reducers and syncing from store to local state is always a pain (and an antipattern that React calls out in the documentation). Shouldn't this state be moved to the store already?

File: packages/compass-telemetry/src/growth-experiments.ts (hunk 4 above)
Comment (jcobis) [reply packages/compass-telemetry/src/growth-experiments.ts:9]: Very good point, I tend to agree. I was following the existing pattern here and I think it might have something to do with how experiments can be iterated on, as in these values can actually change. These values have to correspond to Pholiota values, so for eg. `MOCK_DATA_GENERATOR_20251001` can become `MOCK_DATA_GENERATOR_20260123` if an experiment iteration is conducted. See the [MMS code](https://github.com/10gen/mms/blob/0287f08cb48dd44e5eb19af32bcf31b516447aea/client/packages/types/abTest.ts#L34) as well. I am open to changing this though, particularly for `ExperimentTestGroup`

File: packages/compass-collection/src/components/collection-header/collection-header.tsx (hunk 9 above)
Comment (jcobis) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx]: Will move this to the redux store!

File: packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx (hunk 5 above)
Comment (jcobis) [reply packages/compass-collection/src/components/collection-header-actions/collection-header-actions.spec.tsx]: Consolidated

File: packages/compass-collection/src/components/collection-header/collection-header.tsx [hunk 10]
```diff
@@ -170,8 +187,40 @@ export const CollectionHeader: React.FunctionComponent<
...
+  | 'currentStep'
+  | 'onOpenMockDataModal'
+  | 'onCloseMockDataModal'
+  | 'onSetMockDataGeneratorStep'
+>;
+
+const ConnectedCollectionHeader = connect(mapStateToProps, {
+  onOpenMockDataModal: openMockDataGeneratorModal,
+  onCloseMockDataModal: closeMockDataGeneratorModal,
+  onSetMockDataGeneratorStep: setMockDataGeneratorStep,
```
Comment (ncarbon) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx]: Are these dispatching the actions? I can't tell if these should be the functions exported as `actions` in the store.

Comment (jcobis) [reply packages/compass-collection/src/components/collection-header/collection-header.tsx]: I think so. `connect` dispatches the actions

File: packages/compass-collection/src/stores/collection-tab.ts [hunk 11]
```diff
@@ -130,5 +143,13 @@ export function activatePlugin(
...
     store,
     deactivate: cleanup,
+    actions: {
+      openMockDataGeneratorModal: () =>
+        store.dispatch(openMockDataGeneratorModal()),
+      closeMockDataGeneratorModal: () =>
+        store.dispatch(closeMockDataGeneratorModal()),
+      setMockDataGeneratorStep: (step: MockDataGeneratorStep) =>
+        store.dispatch(setMockDataGeneratorStep(step)),
+    },
```
Comment (gribnoysup) [inline packages/compass-collection/src/stores/collection-tab.ts]: Can you explain what are you doing here? We're passing actions like that only in very old pieces of code that use reflux, not redux

File: packages/compass-collection/src/modules/collection-tab.ts [hunk 12]
```diff
@@ -31,23 +32,47 @@ export type CollectionState = {
...
+}
+
+interface CloseMockDataGeneratorModalAction {
+  type: CollectionActions.CloseMockDataGeneratorModal;
+}
+
+interface SetMockDataGeneratorStepAction {
+  type: CollectionActions.SetMockDataGeneratorStep;
+  step: MockDataGeneratorStep;
+}
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/collection-tab.ts]: Straight up setter actions are an antipattern in redux (I already mentioned this in another PR), actions should be modeled around [actual events happening in the application](https://redux.js.org/style-guide#model-actions-as-events-not-setters)

Comment (jcobis) [reply packages/compass-collection/src/modules/collection-tab.ts]: Thank you, renamed!

File: packages/compass-collection/src/stores/collection-tab.ts (hunk 11 above)
Comment (jcobis) [reply packages/compass-collection/src/stores/collection-tab.ts]: I see, removed!

File: packages/compass-collection/src/components/collection-header/collection-header.tsx [hunk 13]
```diff
@@ -170,8 +187,40 @@ export const CollectionHeader: React.FunctionComponent<
...
+
+const ConnectedCollectionHeader = connect(mapStateToProps, {
+  onOpenMockDataModal: mockDataGeneratorModalOpened,
+  onCloseMockDataModal: mockDataGeneratorModalClosed,
+  onSetMockDataGeneratorStep: mockDataGeneratorStepChanged,
+})(CollectionHeader) as React.ComponentType<ConnectedCollectionHeaderProps>;
+
+export {
+  ConnectedCollectionHeader as CollectionHeader,
+  CollectionHeader as UnconnectedCollectionHeader, // For tests
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx]: Redux documentation is not recommending to test unconnected components when most of their logic relies on store and actions, those tests are redundant and don't really bring much value, we are trying to stick to this rule, please avoid doing that for the future

File: packages/compass-collection/src/components/collection-header/collection-header.tsx [hunk 14]
```diff
@@ -170,8 +187,40 @@ export const CollectionHeader: React.FunctionComponent<
...
+  | 'onOpenMockDataModal'
+  | 'onCloseMockDataModal'
+  | 'onSetMockDataGeneratorStep'
+>;
+
+const ConnectedCollectionHeader = connect(mapStateToProps, {
+  onOpenMockDataModal: mockDataGeneratorModalOpened,
+  onCloseMockDataModal: mockDataGeneratorModalClosed,
+  onSetMockDataGeneratorStep: mockDataGeneratorStepChanged,
+})(CollectionHeader) as React.ComponentType<ConnectedCollectionHeaderProps>;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx]: Redux connect is pretty good at correctly deriving the types for connected components, usually you don't need to assert types unless your types are incorrect, can you clarify why you are doing an assertion here?

File: packages/compass-collection/src/components/collection-header/collection-header.tsx [hunk 15]
```diff
@@ -170,8 +187,40 @@ export const CollectionHeader: React.FunctionComponent<
...
           sourcePipeline={sourcePipeline}
+          onOpenMockDataModal={onOpenMockDataModal}
         />
       </div>
+      <MockDataGeneratorModal
+        isOpen={isMockDataModalOpen}
+        onClose={onCloseMockDataModal}
+        currentStep={currentStep}
+        onCurrentStepChange={onSetMockDataGeneratorStep}
+      />
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/collection-header/collection-header.tsx]: A best practice in Redux is to prefer more granular connections, as you are just passing all this state / actions here without anything in the collection header meaningfully using them, it would be way better to just connect your modal component directly

File: packages/compass-collection/src/modules/collection-tab.ts [hunk 16]
```diff
@@ -62,6 +87,54 @@ const reducer: Reducer<CollectionState, Action> = (
...
+    isAction<MockDataGeneratorStepChangedAction>(
+      action,
+      CollectionActions.MockDataGeneratorStepChanged
+    )
+  ) {
+    return {
+      ...state,
+      mockDataGenerator: {
+        ...state.mockDataGenerator,
+        currentStep: action.step,
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/collection-tab.ts]: I mentioned this in a previous comment: [using actions as setters is an antipattern](https://redux.js.org/style-guide#model-actions-as-events-not-setters) and in most cases reducer should contain [as much of state transitioning logic](https://redux.js.org/style-guide#put-as-much-logic-as-possible-in-reducers) as possible. You renamed the action, but it's still just directly setting the state in the store.

In practical terms, now every place that might call this action also needs to deal with the logic of figuring out how to correctly transition to a next step, worst case scenario this logic is not accounted for, accounted for incorrectly, or drifts between callsites, and someone gets the form in a completely unexpected state. This is why the documentation strongly recommends having logic like that in the reducer: it makes sure that there's only one place that contains the state transitioning logic, if it needs to be changed or amended, there is no question about where to look for it.

Comment (gribnoysup) [reply packages/compass-collection/src/modules/collection-tab.ts]: As a side-note: packing it all inside a single shared action creator is also an option, but still has a downside of someone just dispatching the "raw" action directly and not using correct action creator function, with the logic in reducer it's vitrually impossible to mess up. I thought I'd mention it anyway because there are, indeed, cases where it's just impossible to move the logic right to the reducer (but it's somewhat rare and I don't think that's the case here)

**Comment by jcobis** [comment]
* **#7181** <a href="https://app.graphite.dev/github/pr/mongodb-js/compass/7181?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#7204](https://github.com/mongodb-js/compass/pull/7204) <a href="https://app.graphite.dev/github/pr/mongodb-js/compass/7204?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#7205](https://github.com/mongodb-js/compass/pull/7205) <a href="https://app.graphite.dev/github/pr/mongodb-js/compass/7205?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.dev/github/pr/mongodb-js/compass/7181?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/compass-collection/src/modules/collection-tab.ts [hunk 17]
```diff
@@ -194,6 +223,113 @@ const reducer: Reducer<CollectionState, Action> = (
...
+        nextStep = MockDataGeneratorStep.DOCUMENT_COUNT;
+        break;
+      case MockDataGeneratorStep.DOCUMENT_COUNT:
+        nextStep = MockDataGeneratorStep.PREVIEW_DATA;
+        break;
+      case MockDataGeneratorStep.PREVIEW_DATA:
+        nextStep = MockDataGeneratorStep.GENERATE_DATA;
+        break;
+      default:
+        nextStep = currentStep; // Stay on current step if at end
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/collection-tab.ts:283]: The state of modal being opened is also part of the redx store, if next step is closing the modal, reducer might as well control it

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts [hunk 18]
```diff
@@ -1,41 +0,0 @@
-import { MockDataGeneratorStep } from './types';
-
-export const getNextStep = (
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/utils.ts:3]: Just FYI, moving this logic into a reducer doesn't need to literally mean moving all this code into a reducer, having a clear, separate function to handle this transition make the code cleaner

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.tsx [hunk 19]
```diff
@@ -30,35 +36,30 @@ const rightButtonsStyles = css`
...
-  const onBack = () => {
-    const previousStep = getPreviousStep(currentStep);
-    onCurrentStepChange(previousStep);
-  };
-
-  const onCancel = () => {
-    onOpenChange(false);
-  };
+  const handleNextClick =
+    currentStep === MockDataGeneratorStep.GENERATE_DATA ? onClose : onNextStep;
```
Comment (gribnoysup) [inline packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.tsx:53]: I don't think you need to make this distinction here and mix close clicked action with next step clicked action (being able to clearly distinguish between the two clearly might be just generally useful), you're store is perfectly capable of doing it itself

File: packages/compass-collection/src/modules/collection-tab.ts (hunk 17 above)
Comment (jcobis) [reply packages/compass-collection/src/modules/collection-tab.ts:283]: Yes agreed, planning to do that when we implement that step

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts (hunk 18 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts:3]: I think it makes sense to have this switch statement here because we're going to need to do different things conditionally based on the current step other than just getting the next step, such as kicking off different async actions

File: packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.tsx (hunk 19 above)
Comment (jcobis) [reply packages/compass-collection/src/components/mock-data-generator-modal/mock-data-generator-modal.tsx:53]: Will adjust

File: packages/compass-collection/src/components/mock-data-generator-modal/utils.ts (hunk 18 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/mock-data-generator-modal/utils.ts:3]: Kicking off different async actions doesn't happen in reducers, reducers should never have side-effects
